File: ios/BetterLise/Features/Simulator/CoefficientInput.swift
import Foundation

/// Free-form coefficients (e.g. 1,33): Savoir's weights are rarely multiples of 0,5.
enum CoefficientInput {
    /// The API refuses shared coefficients above this.
    static let maximum = 100.0
    static let step = 0.5

    /// Accepts a comma or a dot as decimal separator; nil when not a usable coefficient.
    static func parse(_ text: String) -> Double? {
        let normalized = text.trimmingCharacters(in: .whitespaces).replacingOccurrences(of: ",", with: ".")
        guard let value = Double(normalized), isValid(value) else { return nil }
        return value
    }

    static func format(_ value: Double) -> String {
        GradeFormat.note(value)
    }

    static func isValid(_ value: Double) -> Bool {
        value.isFinite && value > 0 && value <= maximum
    }

    /// − / + buttons: moves by `step`, rounded to hundredths, never reaching 0 or passing the maximum.
    static func stepped(_ value: Double?, by delta: Double) -> Double {
        let current = value ?? 1
        let next = ((current + delta) * 100).rounded() / 100
        return isValid(next) ? next : current
    }
}

File: android/app/src/main/java/com/betterlise/app/domain/CoefficientInput.kt
package com.betterlise.app.domain

import kotlin.math.roundToLong

/** Free-form coefficients (e.g. 1,33): Savoir's weights are rarely multiples of 0,5. */
object CoefficientInput {
    /** The API refuses shared coefficients above this. */
    const val MAXIMUM = 100.0
    const val STEP = 0.5

    /** Accepts a comma or a dot as decimal separator; null when not a usable coefficient. */
    fun parse(text: String): Double? = text.trim().replace(',', '.').toDoubleOrNull()?.takeIf(::isValid)

    fun isValid(value: Double): Boolean = value.isFinite() && value > 0 && value <= MAXIMUM

    /** − / + buttons: moves by [STEP], rounded to hundredths, never reaching 0 or passing the maximum. */
    fun stepped(value: Double?, delta: Double): Double {
        val current = value ?: 1.0
        val next = ((current + delta) * 100).roundToLong() / 100.0
        return if (isValid(next)) next else current
    }
}
